working/augmentation-libraries/parts_with_imports.md

```diff
@@ -0,0 +1,367 @@
+placed in the header section of the library file after `library`, `import` and
+`export` declarations. Each part file must start with a `part of` directive
+having one of the forms `part of 'library_file_name.dart';` or
+`part of library.name;`, where the library name is the name declared by the
+library file using a `library library.name;` directive. The part file
+designating its containing library is intended to ensure that a part file can
+only ever be part of one library file, which is essential for, for example,
+having useful language support when editing the part file.
+
+The (URI) string version is the most useful for making analysis of a part file
+possible and unique, because it uniquely defines the library by the URI that
+the language itself uses to identify files and libraries. It’s technically
+possible to have two separate libraries with the same declared library name,
+which both include the same part file. The language has a restriction against
+having two libraries with the same declared name in the same program *mainly*
+to avoid this particular issue, but that still makes offline analysis of the
+part file a problem.
+
+Pre-feature part files inherit the entire import scope from the library file.
+Each declaration of the library file and each part file is included in the
+library’s declaration scope. It’s viable to think of part files as being
+textually included in the library file. There is a is even a rule against
+declaring a `part` inclusion of the same file more than once, which matches
+perfectly with that way of thinking.
+
+## Feature
+
+This feature allows a part file to have `import`, `export` and `part`
+directives of its own, where `import` directives only affect the part file
+itself, and its transitive part files. A library is defined by the source code
+of its library file and *all* transitively included part files, which can be an
+arbitrarily deep *tree*. A part file inherits the imports and import prefixes
+of its parent file (the library or part file that included it) into its
+top-level scope, but can choose to ignore or shadow those using its own
+imports.
+
+The design goals and principles are:
+
+*   *Backwards compatible*: If a part file has no `import`, `export` or `part`
+    directive, it works just like it always has.
+    *   Because of that, it’s always safe to move one or more declarations into
+        a new part file. _(Although augmentation declarations modifies that
+        slightly.)_
+    *   Similarly it’s always possible and safe to combine a part file with no
+        `import`s back into its parent file.
+*   *Library member declarations are library-global*: All top-level
+    declarations in the library file and all transitive part files are equal,
+    and are all in scope in every file. They introduce declarations into the
+    library’s declaration scope, which is the most significant scope in all
+    files of the library. If there is any conflict, top-level declarations win!
```

Comment:
> I am actually not sure this is the correct choice.


> Consider that declarations which are not in the current file are not in the lexical scope

Thta's a crazy idea. I just might work!

They are in scope today, but today there is only one scope in a library. It's the same scope that is the top-level scope for every file of a library. 

What I changed here is to have the import scope be per-file, but then I applied the same declaration scope on top of every import scope. There is no change to that.

If we also inherit the declaration scope, and can extend on top of that, ... how would that work?

- A part file's top-level scope extends the *declaration* scope of its parent, not its import scope.
- It extends with its own import scope and prefix scope (as currently specified) and then with its own declarations *and the declarations of all its sub-parts*, but not the declarations of every file in the library.
    - A library file does need to have access to the declarations of its parts, so it needs to include every declaration in the library in the library file's declaration scope. Anything else would be too breaking.
    - A part file inherits its parent's top-level scope (imports,prefixes,all declarations in all subparts, including this one), but can override anything in there with an import, which can then be overridden by the declarations contained in the part file and all its sub-parts.
- If the part file needs to see any declaration from the parent file, which are also the declarations of its sibling parts and any non-shadowed declarations from the rest of the library, it should avoid shadowing with an import or prefix.

(Hmm. Feels familiar. I'm pretty sure I've tried to specify something like that before. Probably in one of the earlier augmentation issues, or maybe in a document I never published.)

That all seems like it could work. 
And it even has some nice properties. It would mean that any declaration that only exists in a parent or sibling part can be ignored.
A declaration in a sibling will no longer conflict with a prefix in this part.

Depending on how augmenting declarations count, a name may be declared in more than one place. 
So about "shadowing a top level member from an import with an augmentation", that would mean declaring an augmentation with a name which would otherwise resolve to an imported declaration from another library. That doesn't completely make sense if the augmentation itself introduces the name. But maybe it should only ever be the base declaration which introduces the name. 

So not sure that situation _needs_ to be an error, but it _can_ be.
* One can see an augmenting declaration as *referencing* the declaration it augments, and therefore the chosen name should *resolve* to that declaration. If it resolves to something that cannot be augmented, that's just a plain error.
* Or one can see the augmenting declaration as simply having the same *name* as the declaration it augments (that's how it's decided what it augments). We don't resolve the *declared* name of a declaration. It *introduces* a name, it doesn't reference one.

In the latter case, there is no need for the augmented declaration to be in scope. The only question is whether the augmenting declaration introduces the name or not. (But if not, that may actually suggest that the former interpretation is the right one.)

The former case, based on that thought, could mean that augmenting declarations do not introduce a name in the surrounding scope (which now matters because the declaration scope is not every declaration in the library any more), and the name of the declaration *is* a reference to the declaration it augments. You can only refer to declaration's name if you can refer to the *base declaration*, and you can only augment it if you can refer to it.

For *augmenting member declarations*, the name should be resolved relative to the type that is being augmented. A static member or constructor should be resolved against the static scope of the class/mixin/etc that it's inside (including all earlier augmentations, even inside the same scope). An instance member is resolved against the interface of the surrounding type (or whatever makes sense for extensions) also including all earlier augmentations.
So, op problem, those are not lexical scope based. If we can refer to the container declaration, we can augment its members

So, TL;DR:
* The top-level scope of the library file is the same one as today (imports, prefixes, all declarations in the library).
* The top-level scope of a part file *extends* the top-level scope of its parent file with its own imports, prefixes and all non-augmenting declarations in its sub-tree.
* Augmenting top-level declarations do not introduce a name in their surrounding scope, instead the name in the augmenting declaration *refers* to the declaration they want to augment, which must be a declaration of the current library, and must be in scope. (So if they work, they don't need to introduce a name, if they don't, they'd be referring to themselves, which makes no sense.)
* Augmenting namespaced declarations (instance, static, constructor) may or may not introduce a name in the lexical scope. It doesn't hurt, and it allows an augmenting declaration body to safely be recursive, without risking a conflict with an imported name.

SGTM. Let's ask others if they can see any issues.